### PR TITLE
Refactor saber blocks in yaml, update build

### LINKED ISSUE #161: Update saber blocks
Make changes corresponding to https://github.com/JCSDA-internal/mpas-jedi/pull/793.  SABER blocks in yamls are reorganized.  Update to a newer build that includes the corresponding changes in SABER.

Do this after the SABER PR is merged.

### PR DESCRIPTION
**Pull request by jjguerrette**
### Description
The SABER blocks conventions were modified in develop-branch source code.  This PR updates the workflow to use the latest conventions, including using an updated build version, and the spack-stack environment.  Both single and double precision builds are provided.

@metdyn provided the new `environmentJEDI.csh`.

@byoung-joo provided the templates for climatological B yaml formatting in https://github.com/JCSDA-internal/mpas-jedi/pull/793.

### Issue closed

Closes #161 

### Tests completed
 - [x] 3dvar_OIE120km_WarmStart
 - [x] 3denvar_OIE120km_WarmStart
 - [x] 3dvar_OIE120km_ColdStart
 - [x] 3dvar_O30kmIE60km_ColdStart
 - [x] 3denvar_O30kmIE60km_WarmStart
 - [x] eda_OIE120km_WarmStart

### CODE REVIEW

**Comment by metdyn** [comment]
@jjguerrette 
I have an updated file config/environmentJEDI.csh in the 
experiment/spack_stack_Oct_25  branch. It contains intel modules. For completeness, this can be useful. 

I noticed the python and miniconda are not loaded in this file, which I think is good for submitting the job script.  I am not sure about the full workflow environment in order to run the workflow.

**Comment by jjguerrette** [comment]
> @jjguerrette I have an updated file config/environmentJEDI.csh in the experiment/spack_stack_Oct_25 branch. It contains intel modules. For completeness, this can be useful.

Thanks for telling me.  I like your if statement for gnu and intel.  I will merge some of your edits with mine.

> I noticed the python and miniconda are not loaded in this file, which I think is good for submitting the job script. I am not sure about the full workflow environment in order to run the workflow.

Currently we use CISL's conda and python environments.  We could switch to using spack stack's later, but it requires ensuring that all python modules needed in our experiments are available.  That will be a separate effort beyond this PR and is not time critical since the CISL solution still works for now.  It's worth discussing at some point.

**Comment by metdyn** [comment]
@jjguerrette  That sounds great.  Me too, I would not like to use python from spack-stack installation: CISL installation is origin and reliable.  We are in the same line.

**Comment by junmeiban** [comment]
Hi @jjguerrette ,
I try to use this workflow, but get some errors(see /glade/scratch/jban/cylc-run/jban_3denvar_OIE120km_WarmStart_TEST/log/job/20180415T0000-06/LinkExternalAnalysis-120km/01/job.out)
I must miss something. Would you please point me how to fix the errors?
Many thanks,
Junmei

**Comment by jjguerrette** [comment]
> I try to use this workflow, but get some errors(see /glade/scratch/jban/cylc-run/jban_3denvar_OIE120km_WarmStart_TEST/log/job/20180415T0000-06/LinkExternalAnalysis-120km/01/job.out)
I must miss something. Would you please point me how to fix the errors?

Thank you @junmeiban.  I'm looking into it.  The error message says that the `conda/latest` module could not be found.  I wonder if there is a conflict with `environmentJEDI.csh`, although `environmentJEDI.csh` is not sourced by `LinkExternalAnalysis.csh`.  Will let you know what I find.

**Comment by jjguerrette** [comment]
@junmeiban, do you have the python module loaded in the terminal login from which you are launching the suite? 

See the error message:
```
Lmod has detected the following error: Cannot load module "conda/latest"
because these module(s) are loaded:
   python

While processing the following module(s):
    Module fullname  Module Filename
    ---------------  ---------------
    conda/latest     /glade/u/apps/dav/modulefiles/default/idep/conda/latest.lua
```

Run `module list` to check if python is loaded.

The python module needs to be unloaded.  If you source `env-script/cheyenne.csh` or `env-script/cheyenne.sh`, the proper modules should get loaded.  MPAS-Workflow does not work with the CISL python module, which will be deprecated in the future.  It also does not work with the python modules that are distributed with spack-stack.  I have not figured out a workaround yet for the case where somebody has sourced the env-scripts distributed with mpas-bundle and then wants to run the workflow in the same terminal login.  Probably we would need to remove the python module loading in the mpas-bundle scripts.

**Comment by junmeiban** [comment]
Currently Loaded Modules:
  1) ncarenv/1.3   2) gnu/10.1.0   3) ncarcompilers/0.5.0   4) netcdf/4.8.1   5) conda/latest   6) cylc/7.8.3   7) graphviz/2.40.1   8) git/2.33.1

**Comment by jjguerrette** [comment]
Do you load the python module in `~/.tcshrc`, `~/.cshrc`, `~/.bash_profile`, or `~/.bashrc`?  Doing so might also cause problems.

**Comment by junmeiban** [comment]
I load python in .tcshrc. I will remove it

**Comment by junmeiban** [comment]
Many thanks @jjguerrette . It works.

**Comment by jjguerrette** [comment]
No problem @junmeiban.  Sorry for the confusion and challenging environment setup.  It needs to be cleaned up at some point.

**Comment by metdyn** [comment]
> Probably we would need to remove the python module loading in the mpas-bundle scripts.

@jjguerrette    I think "MPAS-Workflow/env-setup/cheyenne.csh cheyenne.sh" needs `module reset` before `source /etc/profile.d/modules.csh`.  We can choose not to load python when using the spack-stack, but doing so may fail ioda-bundle (I think).   We need to make sure group members know about this issue.

**Comment by jjguerrette** [comment]
> I think "MPAS-Workflow/env-setup/cheyenne.csh cheyenne.sh" needs `module reset`

What is the difference between `module purge` and `module reset`?  Currently we use `module purge` after `source /etc/profile.d/modules.csh`.

I think the issue that Junmei found is in the beginning of a job submission when cylc sets up the environment for each cylc task.  It is using the user's `.tcshrc` file, regardless of the behavior in `MPAS-Workflow/env-setup/cheyenne.csh` and `cheyenne.sh`.  One workaround would be to add a `module unload python` or `module purge` in `include/tasks.rc` after line 11.

Or use `source env-script/cheyenne.sh` instead of 
```
source /etc/profile.d/modules.sh
module load conda/latest
conda activate npl
```

**Comment by jjguerrette** [comment]
This PR is ready for review now.  All tests passed for me.

**Comment by jjguerrette** [comment]
a5919d6b181bf938ace28292a7ccdce896f37e47 may have broken this feature branch.  I'm checking.

File: config/environmentJEDI.csh [hunk 1]
```diff
@@ -3,25 +3,44 @@
...
 setenv MODULEPATH_ROOT /glade/work/jedipara/cheyenne/spack-stack/modulefiles
 module use /glade/work/jedipara/cheyenne/spack-stack/modulefiles/compilers
 module use /glade/work/jedipara/cheyenne/spack-stack/modulefiles/misc
 
-limit stacksize unlimited
-module use /glade/work/jedipara/cheyenne/spack-stack/spack-stack-v1/envs/skylab-2.0.0-gnu-10.1.0/install/modulefiles/Core
+
+if ( $BuildCompiler =~  *"gnu"* ) then
+#ulimit -s unlimited
+module use /glade/work/jedipara/cheyenne/spack-stack/spack-stack-v1/envs/skylab-1.0.0-gnu-10.1.0/install/modulefiles/Core
```
Comment (jjguerrette) [inline config/environmentJEDI.csh]: I think the problem is with this line.  I used the `env-scripts` in mpas-bundle to build the code.  Those currently use `skylab-2.0.0`, not `skylab-1.0.0`.  Originally I had `skylab-2.0.0` on this line. 
 @metdyn, which is the correct spack-stack version to use?

**Review by junmeiban — APPROVED** [review]

File: config/environmentJEDI.csh (hunk 1 above)
Comment (metdyn) [reply config/environmentJEDI.csh]: Sorry. I was in a hurry for an appointment at that time. I should have pointed this out. I was intended to let you know the if statement for gnu and intel are useful.

Comment (jjguerrette) [reply config/environmentJEDI.csh]: OK, then I will assume skylab-2.0 is correct.  Thank you for confirming.

**Comment by metdyn** [comment]
> What is the difference between `module purge` and `module reset`? Currently we use `module purge` after `source /etc/profile.d/modules.csh`.

Sorry, I did not check in advance on Cheyenne.  `module reset` does not work at all after spack-stack is sourced/loaded. Theoretically `module reset` is described here (https://lmod.readthedocs.io/en/latest/010_user.html). CISL staff mentioned it to me. But it simply does not work.   `module purge` seems better. 

> I think the issue that Junmei found is in the beginning of a job submission when cylc sets up the environment for each cylc task. It is using the user's `.tcshrc` file, regardless of the behavior in `MPAS-Workflow/env-setup/cheyenne.csh` and `cheyenne.sh`. One workaround would be to add a `module unload python` or `module purge` in `include/tasks.rc` after line 11.

You are very right. Junmei's error message said
```
Lmod has detected the following error: Cannot load module "conda/latest"
because these module(s) are loaded:
   python
```
I think `module unload python` can be used, even if `python` is not loaded, the LMOD will give a note and it should not cause an error. Let me check out your branch with added `module unload python`, meanwhile with and without any python module in my .bashrc environment. I am not sure how urgent you want to merge the branch. If so, you should merge.

**Comment by jjguerrette** [comment]
Thank you @metdyn for looking into those environment problems.  I think they can be addressed in a separate issue.  Do you mind creating an issue and assigning yourself?

**Comment by metdyn** [comment]
@jjguerrette Yes, I will create an environment issue.

**Pull request merged**